### PR TITLE
ref(notifications): Get Participants by Provider

### PR DESCRIPTION
**Pull request by mgaeta**
Modify `ActivityNotification.get_participants` to understand notification settings by provider.

### CODE REVIEW

File: src/sentry/notifications/activity/base.py [hunk 1]
```diff
@@ -34,27 +45,46 @@ def _get_subject_prefix(self) -> str:
...
-        if self.activity.user is not None and self.activity.user in participants:
-            receive_own_activity = (
-                UserOption.objects.get_value(
-                    user=self.activity.user, key="self_notifications", default="0"
-                )
-                == "1"
+        participants_by_provider = GroupSubscription.objects.get_participants(self.group)
+        user_option = self.activity.user
+        if user_option:
+            # Remove the actor that created the activity from the recipients list.
```
Comment (ceorourke) [inline src/sentry/notifications/activity/base.py:80]: what about when people opt into receiving notifications about their own activity?

File: src/sentry/notifications/activity/base.py [hunk 2]
```diff
@@ -229,35 +287,33 @@ def send(self) -> None:
...
-                        "reason": GroupSubscriptionReason.descriptions.get(
-                            reason, "are subscribed to this issue"
-                        ),
-                        "unsubscribe_link": generate_signed_link(
-                            user.id,
-                            "sentry-account-email-unsubscribe-issue",
-                            kwargs={"issue_id": group.id},
-                        ),
-                    }
+        for provider, participants in participants_by_provider.items():
```
[the remark this answers is not in the document]
Comment (ceorourke) [reply src/sentry/notifications/activity/base.py:290]: ^ agree here, we do think we'll expand in the future but I think this is okay for now.

File: src/sentry/notifications/helpers.py [hunk 3]
```diff
@@ -95,6 +95,33 @@ def where_should_be_participating(
     ]
 
 
+def get_values_by_provider(
```
Comment (ceorourke) [inline src/sentry/notifications/helpers.py:98]: what about calling this `get_deploy_values_by_provider` to be more specific?

File: src/sentry/notifications/activity/base.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/notifications/activity/base.py]: I updated the comment with a clarification.

File: src/sentry/notifications/activity/base.py [hunk 4]
```diff
@@ -34,27 +45,46 @@ def _get_subject_prefix(self) -> str:
...
+            provider
+            for provider, participants in participants_by_provider.items()
+            if user in participants
+        }
+        if (
+            providers
+            and UserOption.objects.get_value(user=user, key="self_notifications", default="0")
+            == "1"
+        ):
+            return providers
```
[the remark this answers is not in the document]
Comment (mgaeta) [reply src/sentry/notifications/activity/base.py]: Great catch!
I've added a test that should capture this.

File: src/sentry/integrations/slack/notifications.py [hunk 5]
```diff
@@ -1,20 +1,23 @@
...
+    notification: ActivityNotification, user: User, context: Mapping[str, Any]
 ) -> None:
-    pass
+    integrations = Integration.objects.filter(
+        organizations__in=[notification.organization],
+        provider=ExternalProviders.SLACK.name.lower(),
+        status=ObjectStatus.VISIBLE,
+    )
+    for integration in integrations:
+        pass
```
Comment (ceorourke) [inline src/sentry/integrations/slack/notifications.py:23]: will this need a line like:

```ActivityNotification.register(ExternalProviders.SLACK, send_notification_as_slack)```

?

File: src/sentry/notifications/activity/base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgaeta) [reply src/sentry/notifications/activity/base.py:291]: I copied the registry pattern from serializers.
```
for provider, participants in participants_by_provider.items():
	...
    registry[provider](self, user, user_context)
```

**Review by ceorourke — APPROVED** [review]

**Pull request merged**